library/pod-security-policy/allow-privilege-escalation/src.rego

```diff
@@ -0,0 +1,25 @@
+package k8spspallowprivilegeescalationcontainer
+
+violation[{"msg": msg, "details": {}}] {
+    c := input_containers[_]
+    input_allow_priviledge_escalation_allowed(c)
+    msg := sprintf("Privilege escalation container is not allowed: %v", [c.name])
+}
+
+input_allow_priviledge_escalation_allowed(c) {
+    not has_field(c, "securityContext")
+}
```

Comment:
Not sure what the k8s default behavior is when there is no securityContext.  I would have thought the default is that privilegeEscalation is prohibited unless there's an explicit securityContext.allowPrivilegeEscalation.  Couldn't seem to find the answer in the docs.  

But did find in the docs that there are securityContexts at the container level and at the spec level.  The spec level overrides the container level.  I expect we'd need to check both places and do the right thing when it exists at both.

https://kubernetes.io/docs/tasks/configure-pod-container/security-context/

Another question here is: k8s might add the securityContext.allowPrivilegeEscalation on each container before it hands it over to admission control, and so there's no need to check the spec-level and container-level locations.

A methodological question: were you basing the logic here on the PSP implementation and so mirroring exactly what they do?

Reply:
From https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.15/, it seems `allowPrivilegeEscalation` only appears in `Container` spec and `PodSecurityPolicy` spec. So I think checking at the container level should be sufficient. LMK if you see it appearing in `Pod` spec as well. 

> were you basing the logic here on the PSP implementation and so mirroring exactly what they do?

Currently we are mapping each PSP configuration to a Gatekeeper CT as you see in `library/pod-security-policy/README.md`. I hope as we continue to make more progress here, we could combine some of them into one and/or create new ones.

Reply:
I see.  The SecurityContext at the Pod level doesn't allow the allowPrivilegeEscalation. 
 It allows seemingly a subset of the SecurityContext at the Container level.